catkin_ws/src/vision/src/object_detection_utils.py

```diff
@@ -132,20 +132,37 @@ def clean_depth_error(depth_img, error_threshold=0.5):
     return depth_img
 
 def gate_depth(depth_cropped):
-    rows, _ = depth_cropped.shape
-    left_half, right_half = depth_cropped[:int(rows/2), :], depth_cropped[int(rows/2):, :]
-    left_closest = np.min(left_half)
-    right_closest = np.min(right_half)
-    return (left_closest + right_closest) / 2
+    """ divide the picture in two (left and right) - get the mean of the 3 min value on each side -
+    mean the two means """
+    depth_cropped = np.array(depth_cropped)
+    _, cols = depth_cropped.shape
+    left_half, right_half = depth_cropped[:, :int(cols/2)], depth_cropped[:,int(cols/2):]
+    left_flattened = left_half.flatten()
+    left_smallest_values = np.partition(left_flattened, 3)[:3]
+    right_flatten = right_half.flatten()
+    right_smallest_values = np.partition(right_flatten, 3)[:3]
+    return (np.mean(left_smallest_values) + np.mean(right_smallest_values)) / 2
 
 def buoy_depth(depth_cropped):
+    """ get the middle point and the points around it and take the mean 
+        assumption - the buoy is a square object that doesn't have any holes, so taking the middle point
+        should always be where the buoy is """
     rows, cols = depth_cropped.shape
-    middle_point = depth_cropped[int(rows/2), int(cols/2)]
-    return middle_point
+    middle_points = [depth_cropped[int(rows/2), int(cols/2)], depth_cropped[int(rows/2)+1, int(cols/2)],
+                    depth_cropped[int(rows/2)-1, int(cols/2)], depth_cropped[int(rows/2), int(cols/2)+1],
+                    depth_cropped[int(rows/2), int(cols/2)-1]]
+    return np.mean(middle_points)
 
 def lane_marker_depth(depth_cropped):
-    closest_point = np.unravel_index(np.argmin(depth_cropped), depth_cropped.shape)
-    return closest_point
+    """ Get 30 random points and take the mean - (some of them might be the floor but 
```

Comment:
I'm not sure random sampling is the best approach here, it's possible that the lane marker could represent 25% of the image, in which case the 30 points would be more likely to give the depth of the floor than of the lane marker itself. Might be better to filter out the background? I feel like it would be easier and better to filter out the background for all the objects, and then mean the remaining pixels?

Reply:
I think a high pass filter could work for some of the depth finding functions